app/ota_client_stub.py

```diff
@@ -134,3 +215,270 @@ def _find_request(update_request, ecu_id):
+            if e.ecu_id in subecu_directly_connected:
+                subecu_directly_connected[e.ecu_id] = ota_status
+
+            if ota_status == v2.StatusOta.FAILURE:
+                failed_ecu_list.append(e.ecu_id)
+                logger.error(f"Secondary ECU {e.ecu_id} failed: {e.status.status=}")
+            elif ota_status == v2.StatusOta.SUCCESS:
+                success_ecu_list.append(e.ecu_id)
+            elif ota_status == v2.StatusOta.UPDATING:
+                on_going_ecu_list.append(e.ecu_id)
+
+        logger.debug(
+            "\nstatus pulling for all child ecu: \n"
+            f"{failed_ecu_list=}\n"
+            f"{on_going_ecu_list=}\n"
+            f"{success_ecu_list=}"
+        )
+
+        # ensure directly connect ecu status
+        failed_directly_connected_ecu = []
+
+        def is_keep_pulling():
+            for ecu, st in subecu_directly_connected.items():
+                if st == v2.StatusOta.UPDATING:
+                    return True
+                elif st == v2.StatusOta.FAILURE or st == v2.FAILURE:
+                    failed_directly_connected_ecu.append(ecu)
+            return False
+
+        keep_pulling = is_keep_pulling()
+
+        if not keep_pulling:
+            # all directly connected subECUs are in SUCCESS or FAILURE status
+            if failed_directly_connected_ecu:
+                logger.warning(
+                    "all directly connected subecus have finished the ota update,"
+                    "but some subECUs failed to apply the ota update."
+                    f"failed directly subECUs presented: {failed_directly_connected_ecu}"
+                )
+                raise OtaErrorRecoverable(
+                    f"failed directly subECUs presented: {failed_directly_connected_ecu}"
+                )
+            else:
+                logger.info(
+                    "all directly connected secondary ecus are updated successfully."
+                )
+        return True, keep_pulling
+
+    async def _loop_pulling_subecu_status(self):
+        """
+        loop pulling subECUs' status recursively, until all the subECUs are in certain status
+        """
+
+        if not self._ecu_info.get_secondary_ecus():
+            return  # return when no subecu is attached
+
+        while True:
+            # pulling interval
+            await asyncio.sleep(server_cfg.LOOP_QUERYING_SUBECU_STATUS_INTERVAL)
+
+            failed_ecu_list = []
+            all_subecu_reachable, keep_pulling = await self._pulling_subecu_status(
+                failed_ecu_list
+            )
+
+            if all_subecu_reachable:
+                if not keep_pulling:
+                    return
+            else:
+                # unreachable directly connected subECUs presented
+                logger.info(
+                    f"retry pulling subECUs status due to unreachable subECUs {failed_ecu_list}"
+                )
+
+                """
+                if retry_count > retry:
```

Comment:
retry counter is removed since we can't define how long it will take for subecu to reboot for now.